tests/test_map_container_iterator.cpp

```diff
@@ -0,0 +1,347 @@
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .enumerate()
+                      .collect<ftl::map<std::size_t, std::pair<const int, std::string>>>();
+
+  REQUIRE(mapped_map
+          == ftl::map<std::size_t, std::pair<const int, std::string>>{
+            { 0, { 1, "redabc" } },
+            { 1, { 2, "greenabc" } },
+            { 2, { 3, "blueabc" } },
+          });
+}
+
+TEST_CASE(TEST_TAG "map enumerate collect const", TEST_TAG) {
+  const ftl::map<int, std::string> map = { { 1, "red" }, { 2, "green" }, { 3, "blue" } };
+
+  for (const auto &e : map.iter().enumerate()) { INFO(std::get<0>(e)); }
+
+  auto mapped_map = map.iter()
+                      .map([](const auto &entry) {
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .enumerate()
+                      .collect<ftl::map<std::size_t, std::pair<const int, std::string>>>();
+
+  REQUIRE(mapped_map
+          == ftl::map<std::size_t, std::pair<const int, std::string>>{
+            { 0, { 1, "redabc" } },
+            { 1, { 2, "greenabc" } },
+            { 2, { 3, "blueabc" } },
+          });
+}
+
+TEST_CASE(TEST_TAG "map inspect map collect", TEST_TAG) {
+  ftl::map<int, std::string> map = { { 1, "red" }, { 2, "green" }, { 3, "blue" } };
+
+  auto mapped_map = map.iter()
+                      .map([](const auto &entry) {
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .inspect([](const auto &entry) { INFO(entry.second); })
+                      .map([](const auto &entry) {
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .collect<ftl::map<int, std::string>>();
+
+  REQUIRE(mapped_map == ftl::map<int, std::string>{ { 1, "redabcabc" }, { 2, "greenabcabc" }, { 3, "blueabcabc" } });
+}
+
+TEST_CASE(TEST_TAG "map inspect map collect const", TEST_TAG) {
+  const ftl::map<int, std::string> map = { { 1, "red" }, { 2, "green" }, { 3, "blue" } };
+
+  auto mapped_map = map.iter()
+                      .map([](const auto &entry) {
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .inspect([](const auto &entry) { INFO(entry.second); })
+                      .map([](const auto &entry) {
+                        return std::pair<const int, std::string>{ entry.first, entry.second + "abc" };
+                      })
+                      .collect<ftl::map<int, std::string>>();
+
+  REQUIRE(mapped_map == ftl::map<int, std::string>{ { 1, "redabcabc" }, { 2, "greenabcabc" }, { 3, "blueabcabc" } });
+}
+
+// TODO: Fix this code
```

Comment:
?

Reply:
@BIGbadEL This is a bug in map_iterator. operator* should not return value_type as it does now. It should return the same type as the provided mapping function. I would resolve this as another task because this is rather huge change and it will be hard to review it with all of the other changes in this branch.